sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -742,6 +742,23 @@ object InferFiltersFromConstraints extends Rule[LogicalPlan] with PredicateHelpe
  * equivalent [[Literal]] values.
  */
 object ConstantFolding extends Rule[LogicalPlan] {
+  private def isAssociativelyFoldable(e: Expression): Boolean =
```

Comment:
Similar to `ReorderJoin`, we should have a new rule `ReorderAssociativeOperator` to do this optimization, instead of putting it in `ConstantFolding`.